# Task 1 — Data Immersion & Wrangling

## 1. Dataset Loading and Initial Inspection

### Objective

The objective of this step is to load the raw sales dataset and perform an initial inspection of its structure.

We will identify:

- Number of records
- Number of columns
- Column names
- Data types
- Basic dataset information

No data will be modified during this stage.

In [1]:
# Import pandas for reading and analyzing tabular data.

import pandas as pd

In [2]:
# Load the original raw Excel dataset into a pandas DataFrame.
# The raw dataset is kept unchanged for auditability.

df = pd.read_excel("../raw_data/ApexPlanet_DataAnalytics_Dataset.xlsx")

In [3]:
# Display the first five rows of the dataset.
# This gives us a quick look at the structure and values.

df.head()

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
0,ORD100002,2025-02-25,CUST5529,Customer_227,30.0,Female,Bengaluru,Rice,Grocery,7,2829.77,19808.39
1,ORD100003,2025-10-14,CUST3127,Customer_182,63.0,Male,Bengaluru,Book,Education,5,27906.16,139530.80
2,ORD100004,2025-05-13,CUST8887,Customer_487,62.0,Female,Bengaluru,Book,Education,8,37491.06,299928.48
3,ORD100005,2025-12-02,CUST2515,Customer_470,65.0,Female,Kolkata,Mobile,Electronics,9,28541.36,256872.24
4,ORD100006,2025-11-20,CUST4796,Customer_380,44.0,Male,Bengaluru,Rice,Grocery,10,14036.59,140365.90


In [4]:
# Display the number of rows and columns in the dataset.
# The first value represents rows and the second represents columns.

print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 1000
Number of columns: 12


In [5]:
# Display all column names in the dataset.
# This helps us understand the variables available for analysis.

print("Column names:")

for column in df.columns:
    print("-", column)

Column names:
- Order_ID
- Order_Date
- Customer_ID
- Customer_Name
- Age
- Gender
- City
- Product
- Category
- Quantity
- Unit_Price
- Total_Sales


In [6]:
# Display detailed information about the dataset.
# This shows the number of non-missing values and
# the data type of each column.

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Order_ID       1000 non-null   str    
 1   Order_Date     1000 non-null   str    
 2   Customer_ID    1000 non-null   str    
 3   Customer_Name  1000 non-null   str    
 4   Age            980 non-null    float64
 5   Gender         1000 non-null   str    
 6   City           987 non-null    str    
 7   Product        1000 non-null   str    
 8   Category       1000 non-null   str    
 9   Quantity       1000 non-null   int64  
 10  Unit_Price     1000 non-null   float64
 11  Total_Sales    1000 non-null   float64
dtypes: float64(3), int64(1), str(8)
memory usage: 93.9 KB


# 2. Data Dictionary

## Objective

The data dictionary documents the meaning, data type, example value, and business relevance of each column in the dataset.

This helps establish a clear understanding of the dataset before performing data-quality checks and transformations.

## Column Definitions

| Column | Meaning | Data Type | Example | Business Relevance |
|---|---|---|---|---|
| Order_ID | Unique identifier for an order | Text | ORD100002 | Identifies individual sales records |
| Order_Date | Date on which the order was recorded | Date | 2025-02-25 | Enables time-based sales analysis |
| Customer_ID | Identifier for a customer | Text | CUST5529 | Enables customer-level analysis |
| Customer_Name | Name assigned to the customer | Text | Customer_227 | Provides a readable customer reference |
| Age | Age of the customer | Numeric | 30 | Enables demographic analysis |
| Gender | Gender category of the customer | Categorical | Female | Enables demographic comparisons |
| City | City associated with the customer/order | Categorical | Bengaluru | Enables geographic analysis |
| Product | Product purchased | Categorical | Rice | Enables product-level analysis |
| Category | Category of the purchased product | Categorical | Grocery | Enables category-level analysis |
| Quantity | Number of units purchased | Numeric | 7 | Measures sales volume |
| Unit_Price | Price of one unit | Numeric | 2829.77 | Helps analyze pricing and revenue |
| Total_Sales | Total sales value of the transaction | Numeric | 19808.39 | Primary sales-performance measure |

# 3. Initial Data Profiling

## Objective

The purpose of initial data profiling is to understand the structure,
size, data types, and basic characteristics of the dataset before
performing data-quality checks and cleaning.

The following checks will be performed:

- Dataset dimensions
- Column data types
- Non-null values
- Numerical summary statistics
- Number of unique values

In [7]:
# Display the number of rows and columns in the dataset.
# The first value represents the number of rows.
# The second value represents the number of columns.

print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 1000
Number of columns: 12


In [8]:
# Display detailed information about every column.
# This shows:
# - Column names
# - Number of non-missing values
# - Data types
# - Memory usage

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Order_ID       1000 non-null   str    
 1   Order_Date     1000 non-null   str    
 2   Customer_ID    1000 non-null   str    
 3   Customer_Name  1000 non-null   str    
 4   Age            980 non-null    float64
 5   Gender         1000 non-null   str    
 6   City           987 non-null    str    
 7   Product        1000 non-null   str    
 8   Category       1000 non-null   str    
 9   Quantity       1000 non-null   int64  
 10  Unit_Price     1000 non-null   float64
 11  Total_Sales    1000 non-null   float64
dtypes: float64(3), int64(1), str(8)
memory usage: 93.9 KB


In [9]:
# Generate descriptive statistics for numerical columns.
# This provides count, mean, standard deviation,
# minimum, quartiles, and maximum values.

df.describe()

,Age,Quantity,Unit_Price,Total_Sales
count,980.000000,1000.000000,1000.000000,1000.000000
mean,41.360204,5.435000,25486.783410,139399.439650
std,13.822597,2.838632,14179.402361,114100.051546
min,18.000000,1.000000,145.780000,437.340000
25%,30.000000,3.000000,13895.722500,47066.632500
50%,41.000000,5.000000,25398.740000,108594.025000
75%,54.000000,8.000000,37512.382500,203722.882500
max,65.000000,10.000000,49997.530000,493677.500000


In [10]:
# Count the number of unique values in each column.
# This helps us understand identifiers, categories,
# products, cities, and customer information.

unique_values = df.nunique()

print("Unique values in each column:")
print(unique_values)

Unique values in each column:
Order_ID          992
Order_Date        342
Customer_ID       947
Customer_Name     425
Age                48
Gender              2
City                8
Product             6
Category            5
Quantity           10
Unit_Price       1000
Total_Sales      1000
dtype: int64


In [11]:
# Create a compact profiling table for the dataset.
# Each row represents one column in the dataset.

profiling_summary = pd.DataFrame({
    "Data_Type": df.dtypes,
    "Non_Null_Count": df.notna().sum(),
    "Missing_Count": df.isna().sum(),
    "Unique_Values": df.nunique()
})

# Display the profiling summary.
profiling_summary

,Data_Type,Non_Null_Count,Missing_Count,Unique_Values
Order_ID,str,1000,0,992
Order_Date,str,1000,0,342
Customer_ID,str,1000,0,947
Customer_Name,str,1000,0,425
Age,float64,980,20,48
Gender,str,1000,0,2
City,str,987,13,8
Product,str,1000,0,6
Category,str,1000,0,5
Quantity,int64,1000,0,10


# 4. Missing Value Analysis

## Objective

The objective of this step is to identify and measure missing values
in the dataset.

We will:

- Count missing values in each column.
- Calculate the percentage of missing values.
- Identify columns affected by missing values.
- Identify records containing missing values.

No values will be modified or removed at this stage.

In [12]:
# Count the number of missing values in each column.
# isna() identifies missing values.
# sum() counts the missing values column by column.

missing_values = df.isna().sum()

print("Missing values in each column:")
print(missing_values)

Missing values in each column:
Order_ID          0
Order_Date        0
Customer_ID       0
Customer_Name     0
Age              20
Gender            0
City             13
Product           0
Category          0
Quantity          0
Unit_Price        0
Total_Sales       0
dtype: int64


In [13]:
# Calculate the percentage of missing values in each column.
# The missing count is divided by the total number of rows
# and multiplied by 100.

missing_percentage = (df.isna().sum() / len(df)) * 100

print("Missing value percentage:")
print(missing_percentage)

Missing value percentage:
Order_ID         0.0
Order_Date       0.0
Customer_ID      0.0
Customer_Name    0.0
Age              2.0
Gender           0.0
City             1.3
Product          0.0
Category         0.0
Quantity         0.0
Unit_Price       0.0
Total_Sales      0.0
dtype: float64


In [14]:
# Create a summary table containing both the number
# and percentage of missing values for every column.

missing_summary = pd.DataFrame({
    "Missing_Count": df.isna().sum(),
    "Missing_Percentage": (df.isna().sum() / len(df)) * 100
})

# Display the missing-value summary.
missing_summary

,Missing_Count,Missing_Percentage
Order_ID,0,0.0
Order_Date,0,0.0
Customer_ID,0,0.0
Customer_Name,0,0.0
Age,20,2.0
Gender,0,0.0
City,13,1.3
Product,0,0.0
Category,0,0.0
Quantity,0,0.0


In [15]:
# Identify rows containing at least one missing value.
# any(axis=1) checks across each row and returns True
# when at least one column contains a missing value.

rows_with_missing_values = df[df.isna().any(axis=1)]

# Display the affected records.
rows_with_missing_values

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
48,ORD100050,2025-10-03,CUST5523,Customer_432,NaN,Male,Pune,Chair,Furniture,9,30834.87,277513.83
73,ORD100075,2025-03-30,CUST2157,Customer_31,58.0,Male,NaN,Mobile,Electronics,4,21599.26,86397.04
98,ORD100100,2025-11-02,CUST7462,Customer_379,NaN,Male,Patna,Laptop,Electronics,5,553.62,2768.10
148,ORD100150,2025-04-01,CUST4132,Customer_47,NaN,Female,NaN,Chair,Furniture,4,1973.64,7894.56
198,ORD100200,2025-02-24,CUST2953,Customer_435,NaN,Female,Kolkata,Rice,Grocery,8,27608.55,220868.40
223,ORD100225,2025-12-07,CUST1443,Customer_315,37.0,Female,NaN,Rice,Grocery,1,25336.70,25336.70
248,ORD100250,2025-12-20,CUST3169,Customer_138,NaN,Female,Delhi,Mobile,Electronics,7,19829.73,138808.11
298,ORD100300,2025-07-01,CUST7944,Customer_282,NaN,Male,NaN,Chair,Furniture,6,28595.31,171571.86
348,ORD100350,2025-07-11,CUST9709,Customer_409,NaN,Male,Pune,Laptop,Electronics,8,47254.70,378037.60
373,ORD100375,2025-05-06,CUST5835,Customer_362,20.0,Male,NaN,Laptop,Electronics,2,28311.58,56623.16


# 5. Duplicate Analysis

## Objective

The objective of this step is to identify duplicate records
and repeated identifiers in the dataset.

We will check:

- Completely duplicated rows.
- Duplicate Order_ID values.
- Repeated Customer_ID values.
- The number of duplicate records.

No records will be deleted or modified at this stage.

In [16]:
# Check for completely duplicated rows.
# duplicated() identifies rows that are exact copies of
# another row in the dataset.

duplicate_rows = df[df.duplicated()]

# Display the duplicated rows.
duplicate_rows

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales


In [17]:
# Count the number of completely duplicated rows.

print("Number of completely duplicated rows:", len(duplicate_rows))

Number of completely duplicated rows: 0


In [18]:
# Count how many times each Order_ID appears.
# value_counts() returns the frequency of each Order_ID.

order_id_counts = df["Order_ID"].value_counts()

# Show only Order_ID values that appear more than once.

duplicate_order_ids = order_id_counts[order_id_counts > 1]

duplicate_order_ids

Order_ID
ORD100050    9
Name: count, dtype: int64

In [19]:
# Display the total number of Order_ID values
# that appear more than once.

print(
    "Number of duplicate Order_ID values:",
    len(duplicate_order_ids)
)

Number of duplicate Order_ID values: 1


In [20]:
# Identify all records whose Order_ID appears more than once.
# isin() checks whether each Order_ID exists in our duplicate list.

duplicate_order_records = df[
    df["Order_ID"].isin(duplicate_order_ids.index)
]

# Display the records for investigation.
duplicate_order_records

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
48,ORD100050,2025-10-03,CUST5523,Customer_432,NaN,Male,Pune,Chair,Furniture,9,30834.87,277513.83
118,ORD100050,2025-01-22,CUST3194,Customer_114,34.0,Male,Kolkata,Book,Education,2,25795.55,51591.10
238,ORD100050,2025-03-01,CUST8467,Customer_104,23.0,Male,Bengaluru,Book,Education,3,2860.83,8582.49
358,ORD100050,2025-02-11,CUST6467,Customer_181,24.0,Female,Pune,Rice,Grocery,7,45480.39,318362.73
478,ORD100050,2025-11-22,CUST8225,Customer_468,64.0,Male,Bengaluru,Rice,Grocery,8,38957.59,311660.72
598,ORD100050,2025-02-21,CUST8859,Customer_53,NaN,Male,NaN,Laptop,Electronics,7,20144.84,141013.88
718,ORD100050,2025-11-06,CUST7824,Customer_190,31.0,Female,Mumbai,Rice,Grocery,4,42642.59,170570.36
838,ORD100050,2025-08-03,CUST5585,Customer_289,35.0,Male,Kolkata,Chair,Furniture,9,28747.47,258727.23
958,ORD100050,2025-02-09,CUST7188,Customer_89,62.0,Male,Bengaluru,Mobile,Electronics,1,34619.77,34619.77


In [21]:
# Count how many times each Customer_ID appears.

customer_id_counts = df["Customer_ID"].value_counts()

# Show Customer_ID values that appear more than once.

duplicate_customer_ids = customer_id_counts[
    customer_id_counts > 1
]

duplicate_customer_ids

Customer_ID
CUST2515    3
CUST9307    2
CUST3508    2
CUST5806    2
CUST6532    2
CUST3006    2
CUST8053    2
CUST4706    2
CUST3689    2
CUST3001    2
CUST2870    2
CUST4137    2
CUST3366    2
CUST4996    2
CUST4756    2
CUST9645    2
CUST6082    2
CUST6111    2
CUST6345    2
CUST1994    2
CUST7374    2
CUST1463    2
CUST6407    2
CUST9693    2
CUST6067    2
CUST9465    2
CUST3773    2
CUST2794    2
CUST3763    2
CUST4041    2
CUST9469    2
CUST9421    2
CUST9089    2
CUST9942    2
CUST9209    2
CUST9381    2
CUST5363    2
CUST9966    2
CUST6704    2
CUST7568    2
CUST7824    2
CUST7188    2
CUST9458    2
CUST9510    2
CUST6845    2
CUST9081    2
CUST8708    2
CUST7317    2
CUST7026    2
CUST3730    2
CUST9287    2
CUST1997    2
Name: count, dtype: int64

In [22]:
# Display the number of Customer_ID values
# that appear more than once.

print(
    "Number of repeated Customer_ID values:",
    len(duplicate_customer_ids)
)

Number of repeated Customer_ID values: 52


# 6. Data Types & Formatting Analysis

## Objective

The objective of this step is to verify that each column
has an appropriate data type and consistent formatting.

We will:

- Check the current data types.
- Verify date formatting.
- Check numeric columns.
- Check text columns for leading or trailing spaces.
- Identify potential formatting inconsistencies.

No values will be modified at this stage.

In [23]:
# Display the data type of every column.
# This helps us identify columns that may have
# an incorrect or unexpected data type.

print("Data types of all columns:")
print(df.dtypes)

Data types of all columns:
Order_ID             str
Order_Date           str
Customer_ID          str
Customer_Name        str
Age              float64
Gender               str
City                 str
Product              str
Category             str
Quantity           int64
Unit_Price       float64
Total_Sales      float64
dtype: object


In [24]:
# Display a few Order_Date values.
# This allows us to visually inspect the date format.

print("Sample Order_Date values:")
print(df["Order_Date"].head(10))

Sample Order_Date values:
0    2025-02-25
1    2025-10-14
2    2025-05-13
3    2025-12-02
4    2025-11-20
5    2025-12-31
6    2025-10-14
7    2025-12-05
8    2025-11-21
9    2025-04-16
Name: Order_Date, dtype: str


In [25]:
# Check whether Order_Date has a datetime data type.

print(
    "Is Order_Date a datetime column?",
    pd.api.types.is_datetime64_any_dtype(df["Order_Date"])
)

Is Order_Date a datetime column? False


In [26]:
# Define the columns that should contain numeric values.

numeric_columns = [
    "Age",
    "Quantity",
    "Unit_Price",
    "Total_Sales"
]

# Check whether each column contains numeric data.

for column in numeric_columns:
    print(
        column,
        "->",
        pd.api.types.is_numeric_dtype(df[column])
    )

Age -> True
Quantity -> True
Unit_Price -> True
Total_Sales -> True


In [27]:
# Define the columns that contain text values.

text_columns = [
    "Order_ID",
    "Customer_ID",
    "Customer_Name",
    "Gender",
    "City",
    "Product",
    "Category"
]

# Check for leading or trailing spaces in each text column.

for column in text_columns:
    space_issues = (
        df[column].dropna()
        .astype(str)
        .ne(df[column].dropna().astype(str).str.strip())
        .sum()
    )

    print(column, "->", space_issues, "values with extra spaces")

Order_ID -> 0 values with extra spaces
Customer_ID -> 0 values with extra spaces
Customer_Name -> 0 values with extra spaces
Gender -> 0 values with extra spaces
City -> 0 values with extra spaces
Product -> 0 values with extra spaces
Category -> 0 values with extra spaces


In [28]:
# Display the earliest and latest order dates.

print("Earliest Order Date:", df["Order_Date"].min())
print("Latest Order Date:", df["Order_Date"].max())

Earliest Order Date: 2025-01-01
Latest Order Date: 2026-01-01


# 7. Categorical Consistency Analysis

## Objective

The objective of this step is to identify inconsistencies
in categorical columns.

We will:

- Examine unique values in categorical columns.
- Check category frequencies.
- Look for spelling or formatting inconsistencies.
- Verify the relationship between Product and Category.

No values will be modified at this stage.

In [29]:
# Define the categorical columns that we want to investigate.

categorical_columns = [
    "Gender",
    "City",
    "Product",
    "Category"
]

# Display the categorical columns.

print("Categorical columns:")
print(categorical_columns)

Categorical columns:
['Gender', 'City', 'Product', 'Category']


In [30]:
# Display all unique values in each categorical column.
# dropna=False ensures that missing values are also considered.

for column in categorical_columns:
    print("\n", column)
    print(df[column].unique())


 Gender
<StringArray>
['Female', 'Male']
Length: 2, dtype: str

 City
<StringArray>
['Bengaluru',   'Kolkata', 'Hyderabad',     'Patna',    'Mumbai',     'Delhi',
      'Gaya',      'Pune',         nan]
Length: 9, dtype: str

 Product
<StringArray>
['Rice', 'Book', 'Mobile', 'Laptop', 'Shoes', 'Chair']
Length: 6, dtype: str

 Category
<StringArray>
['Grocery', 'Education', 'Electronics', 'Fashion', 'Furniture']
Length: 5, dtype: str


In [31]:
# Count how frequently each category appears.
# This helps us identify unexpected or very rare categories.

for column in categorical_columns:
    print("\n---", column, "---")
    print(df[column].value_counts(dropna=False))


--- Gender ---
Gender
Male      511
Female    489
Name: count, dtype: int64

--- City ---
City
Patna        135
Kolkata      133
Mumbai       131
Hyderabad    125
Delhi        125
Bengaluru    122
Gaya         117
Pune          99
NaN           13
Name: count, dtype: int64

--- Product ---
Product
Mobile    184
Book      178
Laptop    170
Chair     159
Shoes     156
Rice      153
Name: count, dtype: int64

--- Category ---
Category
Electronics    354
Education      178
Furniture      159
Fashion        156
Grocery        153
Name: count, dtype: int64


In [32]:
# Create a cross-tabulation between Product and Category.
# This helps us identify whether a product appears
# under more than one category.

product_category_check = pd.crosstab(
    df["Product"],
    df["Category"]
)

# Display the relationship.

product_category_check

Category,Education,Electronics,Fashion,Furniture,Grocery
Product,,,,,
Book,178,0,0,0,0
Chair,0,0,0,159,0
Laptop,0,170,0,0,0
Mobile,0,184,0,0,0
Rice,0,0,0,0,153
Shoes,0,0,156,0,0


In [33]:
# Count missing values in categorical columns.

categorical_missing = df[categorical_columns].isna().sum()

print("Missing values in categorical columns:")
print(categorical_missing)

Missing values in categorical columns:
Gender       0
City        13
Product      0
Category     0
dtype: int64


# 8. Numerical Values and Business Rule Validation

## Objective

The objective of this step is to verify that numerical values
are logically valid and follow expected business rules.

We will:

- Examine the minimum and maximum values of numerical columns.
- Check for zero and negative values.
- Validate the Age range.
- Validate the Quantity values.
- Verify whether Total_Sales equals Quantity × Unit_Price.

No values will be modified at this stage.

In [34]:
# Define the numerical columns to investigate.

numeric_columns = [
    "Age",
    "Quantity",
    "Unit_Price",
    "Total_Sales"
]

# Display minimum and maximum values for each numerical column.

for column in numeric_columns:
    print(
        column,
        "-> Minimum:", df[column].min(),
        "| Maximum:", df[column].max()
    )

Age -> Minimum: 18.0 | Maximum: 65.0
Quantity -> Minimum: 1 | Maximum: 10
Unit_Price -> Minimum: 145.78 | Maximum: 49997.53
Total_Sales -> Minimum: 437.34 | Maximum: 493677.5


In [35]:
# Count zero and negative values in each numerical column.

for column in numeric_columns:
    zero_count = (df[column] == 0).sum()
    negative_count = (df[column] < 0).sum()

    print(
        column,
        "-> Zero values:", zero_count,
        "| Negative values:", negative_count
    )

Age -> Zero values: 0 | Negative values: 0
Quantity -> Zero values: 0 | Negative values: 0
Unit_Price -> Zero values: 0 | Negative values: 0
Total_Sales -> Zero values: 0 | Negative values: 0


In [36]:
# Display descriptive statistics for Age.
# Missing Age values are automatically ignored.

print(df["Age"].describe())

count    980.000000
mean      41.360204
std       13.822597
min       18.000000
25%       30.000000
50%       41.000000
75%       54.000000
max       65.000000
Name: Age, dtype: float64


In [37]:
# Identify records with potentially invalid Age values.
# Here we check for Age less than or equal to 0
# or greater than 100.

invalid_age = df[
    (df["Age"] <= 0) |
    (df["Age"] > 100)
]

print("Number of potentially invalid Age values:", len(invalid_age))

invalid_age

Number of potentially invalid Age values: 0


,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales


In [38]:
# Display the unique Quantity values in sorted order.
# This helps us understand the allowed quantity range.

print(
    "Unique Quantity values:",
    sorted(df["Quantity"].dropna().unique())
)

Unique Quantity values: [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10)]


In [39]:
# Identify records where Quantity is zero or negative.

invalid_quantity = df[df["Quantity"] <= 0]

print(
    "Number of zero or negative Quantity values:",
    len(invalid_quantity)
)

invalid_quantity

Number of zero or negative Quantity values: 0


,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales


In [40]:
# Calculate the expected Total_Sales using
# Quantity multiplied by Unit_Price.

calculated_sales = df["Quantity"] * df["Unit_Price"]

# Compare the calculated value with the existing Total_Sales.
# round(2) is used to avoid minor floating-point precision issues.

sales_mismatch = df[
    calculated_sales.round(2) != df["Total_Sales"].round(2)
]

print(
    "Number of Total_Sales mismatches:",
    len(sales_mismatch)
)

sales_mismatch

Number of Total_Sales mismatches: 0


,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales


In [41]:
# Create a compact summary of the numerical
# and business-rule validation results.

print("NUMERICAL VALIDATION SUMMARY")
print("-" * 35)

print("Invalid Age values:", len(invalid_age))
print("Invalid Quantity values:", len(invalid_quantity))
print(
    "Negative Unit_Price values:",
    (df["Unit_Price"] < 0).sum()
)
print(
    "Negative Total_Sales values:",
    (df["Total_Sales"] < 0).sum()
)
print(
    "Total_Sales formula mismatches:",
    len(sales_mismatch)
)

NUMERICAL VALIDATION SUMMARY
-----------------------------------
Invalid Age values: 0
Invalid Quantity values: 0
Negative Unit_Price values: 0
Negative Total_Sales values: 0
Total_Sales formula mismatches: 0


# 9. Outlier Detection and Investigation

## Objective

The objective of this step is to identify statistically unusual
numerical observations in the dataset.

We will:

- Use the Interquartile Range (IQR) method.
- Identify potential outliers in numerical columns.
- Count the number of potential outliers.
- Inspect the identified records.

Potential outliers will not be removed automatically.
A separate cleaning decision will be made later.

In [42]:
# Define the numerical columns for outlier analysis.

outlier_columns = [
    "Age",
    "Quantity",
    "Unit_Price",
    "Total_Sales"
]

print("Columns selected for outlier analysis:")
print(outlier_columns)

Columns selected for outlier analysis:
['Age', 'Quantity', 'Unit_Price', 'Total_Sales']


In [43]:
# Calculate the first quartile (Q1), third quartile (Q3),
# and Interquartile Range (IQR) for each numerical column.

outlier_summary = []

for column in outlier_columns:
    q1 = df[column].quantile(0.25)
    q3 = df[column].quantile(0.75)
    iqr = q3 - q1

    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr

    outlier_count = (
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ).sum()

    outlier_summary.append({
        "Column": column,
        "Q1": q1,
        "Q3": q3,
        "IQR": iqr,
        "Lower_Bound": lower_bound,
        "Upper_Bound": upper_bound,
        "Outlier_Count": outlier_count
    })

outlier_summary = pd.DataFrame(outlier_summary)

outlier_summary

,Column,Q1,Q3,IQR,Lower_Bound,Upper_Bound,Outlier_Count
0,Age,30.0000,54.0000,24.00,-6.0000,90.0000,0
1,Quantity,3.0000,8.0000,5.00,-4.5000,15.5000,0
2,Unit_Price,13895.7225,37512.3825,23616.66,-21529.2675,72937.3725,0
3,Total_Sales,47066.6325,203722.8825,156656.25,-187917.7425,438707.2575,19


In [44]:
# Create a dictionary to store the potential outlier records
# for each numerical column.

potential_outliers = {}

for column in outlier_columns:
    q1 = df[column].quantile(0.25)
    q3 = df[column].quantile(0.75)
    iqr = q3 - q1

    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr

    potential_outliers[column] = df[
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ]

# Display the potential outliers for each column.

for column, records in potential_outliers.items():
    print(f"\n--- Potential outliers in {column} ---")
    print(records[[column]].to_string(index=False))


--- Potential outliers in Age ---
Empty DataFrame
Columns: [Age]
Index: []

--- Potential outliers in Quantity ---
Empty DataFrame
Columns: [Quantity]
Index: []

--- Potential outliers in Unit_Price ---
Empty DataFrame
Columns: [Unit_Price]
Index: []

--- Potential outliers in Total_Sales ---
 Total_Sales
   493677.50
   439609.50
   439499.43
   444513.78
   490866.40
   474935.90
   439266.33
   439972.20
   439469.60
   440456.60
   482551.90
   462701.30
   458774.90
   470798.60
   444016.40
   492174.10
   473144.20
   485667.80
   468522.00


In [45]:
# Display the complete records that have potential
# Total_Sales outliers.

total_sales_outliers = potential_outliers["Total_Sales"]

total_sales_outliers

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
50,ORD100052,2025-09-09,CUST2062,Customer_254,58.0,Male,Patna,Book,Education,10,49367.75,493677.50
81,ORD100083,2025-01-06,CUST8346,Customer_384,30.0,Male,Hyderabad,Laptop,Electronics,9,48845.50,439609.50
150,ORD100152,2025-02-08,CUST5188,Customer_286,53.0,Male,Gaya,Laptop,Electronics,9,48833.27,439499.43
173,ORD100175,2025-06-07,CUST3564,Customer_267,41.0,Male,Patna,Rice,Grocery,9,49390.42,444513.78
220,ORD100222,2025-07-13,CUST1711,Customer_36,20.0,Female,Bengaluru,Shoes,Fashion,10,49086.64,490866.40
364,ORD100366,2025-03-13,CUST9551,Customer_230,18.0,Male,Bengaluru,Mobile,Electronics,10,47493.59,474935.90
396,ORD100398,2025-03-07,CUST6600,Customer_242,58.0,Female,Bengaluru,Mobile,Electronics,9,48807.37,439266.33
432,ORD100434,2025-06-10,CUST4982,Customer_370,33.0,Male,Delhi,Mobile,Electronics,10,43997.22,439972.20
448,ORD100450,2025-02-13,CUST6714,Customer_31,NaN,Female,NaN,Book,Education,10,43946.96,439469.60
462,ORD100464,2025-09-17,CUST1710,Customer_355,46.0,Female,Hyderabad,Book,Education,10,44045.66,440456.60


# 10. Cleaning Decision Plan

## Objective

Based on the data-quality investigation, we will decide which
issues require correction and which observations should be retained.

The cleaning decisions will consider:

- Data quality findings.
- Business rules.
- Whether an observation represents an actual error.
- Whether changing or removing data could introduce bias.

No cleaning code will be executed in this step.

## Cleaning Decision Table

| Issue | Finding | Decision | Reason |
|---|---|---|---|
| Missing Age | Missing values identified | Investigate before deciding | Age may be recoverable from reliable information; otherwise an appropriate strategy is required |
| Missing City | Missing values identified | Retain temporarily | Missing location should not be guessed without evidence |
| Duplicate complete rows | No complete duplicate rows identified | No action | No exact duplicate records require removal |
| Duplicate Order_ID | Repeated Order_ID values identified | Investigate before deciding | Repeated IDs may represent duplicate orders or identifier-quality issues |
| Repeated Customer_ID | Repeated Customer_ID values identified | Retain | A customer can legitimately place multiple orders |
| Order_Date format | Date column requires formatting correction | Convert during cleaning | A consistent datetime format is required for reliable analysis |
| Categorical values | No major categorical inconsistency identified | No action | Values appear sufficiently consistent |
| Product–Category relationship | Product/category relationship checked | No action if consistent | Existing business relationship is valid |
| Numerical values | No obvious invalid numerical values identified | No action | Values satisfy the basic numerical/business rules |
| Total_Sales calculation | Formula validation performed | Retain if no mismatches | Existing sales values agree with Quantity × Unit_Price |
| Statistical outliers | Potential outliers identified | Retain unless proven erroneous | Statistical unusualness does not mean the value is incorrect |

In [46]:
# Display records where Age is missing.
# We will inspect Customer_ID, Customer_Name, Gender, City,
# and other available information to determine whether
# the missing Age can be reliably recovered.

missing_age_records = df[df["Age"].isna()]

missing_age_records

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
48,ORD100050,2025-10-03,CUST5523,Customer_432,NaN,Male,Pune,Chair,Furniture,9,30834.87,277513.83
98,ORD100100,2025-11-02,CUST7462,Customer_379,NaN,Male,Patna,Laptop,Electronics,5,553.62,2768.10
148,ORD100150,2025-04-01,CUST4132,Customer_47,NaN,Female,NaN,Chair,Furniture,4,1973.64,7894.56
198,ORD100200,2025-02-24,CUST2953,Customer_435,NaN,Female,Kolkata,Rice,Grocery,8,27608.55,220868.40
248,ORD100250,2025-12-20,CUST3169,Customer_138,NaN,Female,Delhi,Mobile,Electronics,7,19829.73,138808.11
298,ORD100300,2025-07-01,CUST7944,Customer_282,NaN,Male,NaN,Chair,Furniture,6,28595.31,171571.86
348,ORD100350,2025-07-11,CUST9709,Customer_409,NaN,Male,Pune,Laptop,Electronics,8,47254.70,378037.60
398,ORD100400,2025-04-20,CUST5102,Customer_416,NaN,Female,Hyderabad,Book,Education,3,26269.62,78808.86
448,ORD100450,2025-02-13,CUST6714,Customer_31,NaN,Female,NaN,Book,Education,10,43946.96,439469.60
498,ORD100500,2025-06-26,CUST2835,Customer_468,NaN,Male,Kolkata,Chair,Furniture,9,7268.14,65413.26


In [47]:
# Check whether the customers with missing Age
# appear elsewhere in the dataset with a known Age.

missing_age_customers = missing_age_records["Customer_ID"].dropna().unique()

age_reference_records = df[
    df["Customer_ID"].isin(missing_age_customers) &
    df["Age"].notna()
]

age_reference_records[
    ["Customer_ID", "Customer_Name", "Age", "Gender", "City"]
]

,Customer_ID,Customer_Name,Age,Gender,City
983,CUST3730,Customer_397,58.0,Female,Gaya


In [48]:
# Display records where City is missing.

missing_city_records = df[df["City"].isna()]

missing_city_records

,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
73,ORD100075,2025-03-30,CUST2157,Customer_31,58.0,Male,NaN,Mobile,Electronics,4,21599.26,86397.04
148,ORD100150,2025-04-01,CUST4132,Customer_47,NaN,Female,NaN,Chair,Furniture,4,1973.64,7894.56
223,ORD100225,2025-12-07,CUST1443,Customer_315,37.0,Female,NaN,Rice,Grocery,1,25336.70,25336.70
298,ORD100300,2025-07-01,CUST7944,Customer_282,NaN,Male,NaN,Chair,Furniture,6,28595.31,171571.86
373,ORD100375,2025-05-06,CUST5835,Customer_362,20.0,Male,NaN,Laptop,Electronics,2,28311.58,56623.16
448,ORD100450,2025-02-13,CUST6714,Customer_31,NaN,Female,NaN,Book,Education,10,43946.96,439469.60
523,ORD100525,2025-12-31,CUST1090,Customer_316,45.0,Female,NaN,Shoes,Fashion,7,751.97,5263.79
598,ORD100050,2025-02-21,CUST8859,Customer_53,NaN,Male,NaN,Laptop,Electronics,7,20144.84,141013.88
673,ORD100675,2025-11-20,CUST8502,Customer_220,47.0,Male,NaN,Laptop,Electronics,1,14756.18,14756.18
748,ORD100750,2025-07-18,CUST3730,Customer_273,NaN,Female,NaN,Chair,Furniture,10,43808.06,438080.60


In [49]:
# Check whether customers with missing City
# appear elsewhere with a known City.

missing_city_customers = missing_city_records["Customer_ID"].dropna().unique()

city_reference_records = df[
    df["Customer_ID"].isin(missing_city_customers) &
    df["City"].notna()
]

city_reference_records[
    ["Customer_ID", "Customer_Name", "City", "Gender", "Age"]
]

,Customer_ID,Customer_Name,City,Gender,Age
128,CUST2870,Customer_495,Patna,Male,33.0
983,CUST3730,Customer_397,Gaya,Female,58.0


In [50]:
# Display all records associated with repeated Order_ID values.
# This helps us determine whether repeated IDs represent
# genuinely duplicated orders or different records.

duplicate_order_records[
    [
        "Order_ID",
        "Order_Date",
        "Customer_ID",
        "Customer_Name",
        "Product",
        "Category",
        "Quantity",
        "Unit_Price",
        "Total_Sales"
    ]
]

,Order_ID,Order_Date,Customer_ID,Customer_Name,Product,Category,Quantity,Unit_Price,Total_Sales
48,ORD100050,2025-10-03,CUST5523,Customer_432,Chair,Furniture,9,30834.87,277513.83
118,ORD100050,2025-01-22,CUST3194,Customer_114,Book,Education,2,25795.55,51591.10
238,ORD100050,2025-03-01,CUST8467,Customer_104,Book,Education,3,2860.83,8582.49
358,ORD100050,2025-02-11,CUST6467,Customer_181,Rice,Grocery,7,45480.39,318362.73
478,ORD100050,2025-11-22,CUST8225,Customer_468,Rice,Grocery,8,38957.59,311660.72
598,ORD100050,2025-02-21,CUST8859,Customer_53,Laptop,Electronics,7,20144.84,141013.88
718,ORD100050,2025-11-06,CUST7824,Customer_190,Rice,Grocery,4,42642.59,170570.36
838,ORD100050,2025-08-03,CUST5585,Customer_289,Chair,Furniture,9,28747.47,258727.23
958,ORD100050,2025-02-09,CUST7188,Customer_89,Mobile,Electronics,1,34619.77,34619.77


## Final Cleaning Decisions

Based on the investigation performed in Step 10:

### Missing Age

- 20 records have missing Age values.
- One customer, CUST3730, has a reliable Age value of 58 in another record.
- The remaining missing Age values cannot be reliably recovered from the available dataset.
- Therefore, only the reliably recoverable value will be considered for correction.

### Missing City

- 13 records have missing City values.
- CUST3730 has a known City value of Gaya in another record.
- CUST2870 has a known City value of Patna in another record.
- These values can be reliably recovered.
- The remaining missing City values will not be guessed.

### Duplicate Order_ID

- No completely duplicated rows were found.
- ORD100050 appears 9 times.
- These records contain different transaction details and therefore represent different records rather than exact duplicates.
- The repeated Order_ID will not be used as a reason to delete records.
- The Order_ID issue will be handled carefully during the cleaning stage because the correct replacement identifiers cannot be determined with certainty from the available data.

### Order_Date

- The Order_Date column was not recognized as a datetime column during validation.
- It will be converted to a consistent datetime format during the cleaning stage.

### Statistical Outliers

- Potential outliers were identified using the IQR method.
- Outliers will not be removed automatically because statistical unusualness does not prove that a transaction is incorrect.

# 11. Data Cleaning and Transformation

## Objective

The objective of this step is to apply the approved cleaning
decisions from Step 10.

A separate copy of the original dataset will be created so that
the raw dataset remains unchanged.

The cleaning process will:

- Standardize the Order_Date data type.
- Recover only reliably identifiable missing values.
- Preserve valid transaction records.
- Preserve statistical outliers unless proven incorrect.
- Document data-quality limitations rather than inventing values.

In [51]:
# Create a separate copy of the original dataset.
# This protects the raw dataset from accidental modification.

df_cleaned = df.copy()

# Check that the cleaned DataFrame has the same
# number of rows and columns as the original.

print("Original dataset shape:", df.shape)
print("Cleaned dataset shape:", df_cleaned.shape)

Original dataset shape: (1000, 12)
Cleaned dataset shape: (1000, 12)


In [52]:
# Convert Order_Date into Pandas datetime format.
# errors="coerce" converts values that cannot be interpreted
# as dates into missing values instead of causing an error.

df_cleaned["Order_Date"] = pd.to_datetime(
    df_cleaned["Order_Date"],
    errors="coerce"
)

# Verify the resulting data type.

print("Order_Date data type after cleaning:")
print(df_cleaned["Order_Date"].dtype)

Order_Date data type after cleaning:
datetime64[us]


In [53]:
# Check whether any Order_Date values became missing
# during the conversion.

print(
    "Missing Order_Date values after conversion:",
    df_cleaned["Order_Date"].isna().sum()
)

Missing Order_Date values after conversion: 0


In [54]:
# Recover the missing Age for CUST3730.
# The value 58 was verified from another record
# belonging to the same customer.

df_cleaned.loc[
    df_cleaned["Customer_ID"] == "CUST3730",
    "Age"
] = df_cleaned.loc[
    df_cleaned["Customer_ID"] == "CUST3730",
    "Age"
].fillna(58)

# Verify the result.

print(
    "CUST3730 Age values:",
    df_cleaned.loc[
        df_cleaned["Customer_ID"] == "CUST3730",
        "Age"
    ].unique()
)

CUST3730 Age values: [58.]


In [55]:
# Recover the verified City value for CUST3730.
# Another record for this customer showed Gaya.

df_cleaned.loc[
    df_cleaned["Customer_ID"] == "CUST3730",
    "City"
] = df_cleaned.loc[
    df_cleaned["Customer_ID"] == "CUST3730",
    "City"
].fillna("Gaya")


# Recover the verified City value for CUST2870.
# Another record for this customer showed Patna.

df_cleaned.loc[
    df_cleaned["Customer_ID"] == "CUST2870",
    "City"
] = df_cleaned.loc[
    df_cleaned["Customer_ID"] == "CUST2870",
    "City"
].fillna("Patna")

In [56]:
# Display the records for the customers whose missing
# values were recovered.

df_cleaned[
    df_cleaned["Customer_ID"].isin(
        ["CUST3730", "CUST2870"]
    )
][
    [
        "Customer_ID",
        "Customer_Name",
        "Age",
        "City"
    ]
]

,Customer_ID,Customer_Name,Age,City
128,CUST2870,Customer_495,33.0,Patna
748,CUST3730,Customer_273,58.0,Gaya
973,CUST2870,Customer_311,25.0,Patna
983,CUST3730,Customer_397,58.0,Gaya


In [57]:
# Check the remaining missing Age and City values
# after applying only the verified corrections.

print("Remaining missing Age:", df_cleaned["Age"].isna().sum())
print("Remaining missing City:", df_cleaned["City"].isna().sum())

Remaining missing Age: 19
Remaining missing City: 11


## Handling Duplicate Order IDs

The source dataset contains repeated Order_ID values that
represent different transaction records.

The original Order_ID values will be preserved because the
correct replacement identifiers cannot be determined from
the available data.

A separate Transaction_ID will be created as a unique
identifier for each transaction record.

In [58]:
# Identify Order_ID values that occur more than once.
# We are checking the cleaned dataset before creating
# a separate transaction identifier.

order_id_counts_cleaned = df_cleaned["Order_ID"].value_counts()

duplicate_order_ids_cleaned = order_id_counts_cleaned[
    order_id_counts_cleaned > 1
]

print("Duplicate Order_ID values:")
print(duplicate_order_ids_cleaned)

Duplicate Order_ID values:
Order_ID
ORD100050    9
Name: count, dtype: int64


In [59]:
# Create a unique Transaction_ID for every transaction.
# This is a newly generated analysis identifier.
# It does not replace or overwrite the original Order_ID.

df_cleaned.insert(
    0,
    "Transaction_ID",
    [
        f"TXN{i:06d}"
        for i in range(1, len(df_cleaned) + 1)
    ]
)

# Display the first few records.

df_cleaned.head()

,Transaction_ID,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
0,TXN000001,ORD100002,2025-02-25,CUST5529,Customer_227,30.0,Female,Bengaluru,Rice,Grocery,7,2829.77,19808.39
1,TXN000002,ORD100003,2025-10-14,CUST3127,Customer_182,63.0,Male,Bengaluru,Book,Education,5,27906.16,139530.80
2,TXN000003,ORD100004,2025-05-13,CUST8887,Customer_487,62.0,Female,Bengaluru,Book,Education,8,37491.06,299928.48
3,TXN000004,ORD100005,2025-12-02,CUST2515,Customer_470,65.0,Female,Kolkata,Mobile,Electronics,9,28541.36,256872.24
4,TXN000005,ORD100006,2025-11-20,CUST4796,Customer_380,44.0,Male,Bengaluru,Rice,Grocery,10,14036.59,140365.90


In [60]:
# Check whether every Transaction_ID is unique.

transaction_id_count = df_cleaned["Transaction_ID"].nunique()

print("Total rows:", len(df_cleaned))
print("Unique Transaction_IDs:", transaction_id_count)
print(
    "Are all Transaction_IDs unique?",
    transaction_id_count == len(df_cleaned)
)

Total rows: 1000
Unique Transaction_IDs: 1000
Are all Transaction_IDs unique? True


In [61]:
# Confirm that the original Order_ID column still contains
# the source values and has not been replaced.

print("Original Order_ID for affected records:")

df_cleaned[
    df_cleaned["Order_ID"] == "ORD100050"
][
    ["Transaction_ID", "Order_ID", "Customer_ID", "Product", "Total_Sales"]
]

Original Order_ID for affected records:


,Transaction_ID,Order_ID,Customer_ID,Product,Total_Sales
48,TXN000049,ORD100050,CUST5523,Chair,277513.83
118,TXN000119,ORD100050,CUST3194,Book,51591.10
238,TXN000239,ORD100050,CUST8467,Book,8582.49
358,TXN000359,ORD100050,CUST6467,Rice,318362.73
478,TXN000479,ORD100050,CUST8225,Rice,311660.72
598,TXN000599,ORD100050,CUST8859,Laptop,141013.88
718,TXN000719,ORD100050,CUST7824,Rice,170570.36
838,TXN000839,ORD100050,CUST5585,Chair,258727.23
958,TXN000959,ORD100050,CUST7188,Mobile,34619.77


## Handling Unresolved Missing Values

Some missing Age and City values could not be reliably
recovered from the available dataset.

These values will remain missing rather than being
replaced with guessed or unsupported values.

The unresolved missing values will be documented and
handled appropriately during subsequent analysis.

In [62]:
# Check the remaining missing values after
# applying only the verified corrections.

remaining_missing = df_cleaned.isna().sum()

# Display only columns that still contain missing values.

remaining_missing[
    remaining_missing > 0
]

Age     19
City    11
dtype: int64

In [63]:
# Compare missing values before and after cleaning.
# This helps ensure that our cleaning operations
# did not accidentally introduce new missing values.

original_missing = df.isna().sum()
cleaned_missing = df_cleaned.isna().sum()

missing_comparison = pd.DataFrame({
    "Original_Missing": original_missing,
    "Cleaned_Missing": cleaned_missing
})

missing_comparison[
    (missing_comparison["Original_Missing"] > 0) |
    (missing_comparison["Cleaned_Missing"] > 0)
]

,Original_Missing,Cleaned_Missing
Age,20.0,19
City,13.0,11


## 11.8 Final Cleaning Validation

The cleaned dataset will be validated to ensure that:

- The expected number of records is preserved.
- Transaction_ID values are unique.
- Order_Date is stored as a datetime.
- Verified missing values were recovered.
- Remaining missing values are documented.
- Total_Sales still agrees with Quantity × Unit_Price.
- No unexpected negative numerical values were introduced.
- Original Order_ID values remain preserved.

In [64]:
# Check that cleaning did not accidentally add or remove records.

print("Original shape:", df.shape)
print("Cleaned shape:", df_cleaned.shape)

print(
    "Same number of rows:",
    len(df) == len(df_cleaned)
)

Original shape: (1000, 12)
Cleaned shape: (1000, 13)
Same number of rows: True


In [65]:
# Confirm that every transaction has a unique Transaction_ID.

print(
    "All Transaction_IDs unique:",
    df_cleaned["Transaction_ID"].is_unique
)

print(
    "Missing Transaction_IDs:",
    df_cleaned["Transaction_ID"].isna().sum()
)

All Transaction_IDs unique: True
Missing Transaction_IDs: 0


In [66]:
# Confirm that Order_Date is now stored as a datetime column.

print(
    "Order_Date is datetime:",
    pd.api.types.is_datetime64_any_dtype(
        df_cleaned["Order_Date"]
    )
)

print(
    "Missing Order_Date values:",
    df_cleaned["Order_Date"].isna().sum()
)

Order_Date is datetime: True
Missing Order_Date values: 0


In [67]:
# Recalculate Total_Sales from Quantity and Unit_Price
# and compare it with the cleaned dataset.

calculated_sales_cleaned = (
    df_cleaned["Quantity"] *
    df_cleaned["Unit_Price"]
)

sales_mismatch_cleaned = df_cleaned[
    calculated_sales_cleaned.round(2) !=
    df_cleaned["Total_Sales"].round(2)
]

print(
    "Total_Sales mismatches after cleaning:",
    len(sales_mismatch_cleaned)
)

Total_Sales mismatches after cleaning: 0


In [68]:
# Check for negative values after cleaning.

print(
    "Negative Age:",
    (df_cleaned["Age"] < 0).sum()
)

print(
    "Negative Quantity:",
    (df_cleaned["Quantity"] < 0).sum()
)

print(
    "Negative Unit_Price:",
    (df_cleaned["Unit_Price"] < 0).sum()
)

print(
    "Negative Total_Sales:",
    (df_cleaned["Total_Sales"] < 0).sum()
)

Negative Age: 0
Negative Quantity: 0
Negative Unit_Price: 0
Negative Total_Sales: 0


In [69]:
# Display the final missing-value counts
# for the cleaned dataset.

final_missing = df_cleaned.isna().sum()

final_missing[
    final_missing > 0
]

Age     19
City    11
dtype: int64

In [70]:
# Display the first few rows of the cleaned dataset
# to visually verify the final structure.

df_cleaned.head()

,Transaction_ID,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
0,TXN000001,ORD100002,2025-02-25,CUST5529,Customer_227,30.0,Female,Bengaluru,Rice,Grocery,7,2829.77,19808.39
1,TXN000002,ORD100003,2025-10-14,CUST3127,Customer_182,63.0,Male,Bengaluru,Book,Education,5,27906.16,139530.80
2,TXN000003,ORD100004,2025-05-13,CUST8887,Customer_487,62.0,Female,Bengaluru,Book,Education,8,37491.06,299928.48
3,TXN000004,ORD100005,2025-12-02,CUST2515,Customer_470,65.0,Female,Kolkata,Mobile,Electronics,9,28541.36,256872.24
4,TXN000005,ORD100006,2025-11-20,CUST4796,Customer_380,44.0,Male,Bengaluru,Rice,Grocery,10,14036.59,140365.90


# 12. Create Cleaned Dataset

## Objective

The validated cleaned DataFrame will be exported as an
analysis-ready Excel dataset.

The original raw dataset will remain unchanged.

The cleaned dataset contains:

- A unique Transaction_ID for analysis.
- Standardized Order_Date data type.
- Verified recovery of missing Age and City values.
- Original Order_ID values preserved.
- Unresolved missing values retained where reliable information
  was unavailable.
- Statistical outliers retained because they were not proven
  to be data errors.

In [71]:
# Define the output path for the cleaned dataset.
# The file will be saved inside the cleaned_data folder.

output_path = "../cleaned_data/ApexPlanet_Cleaned_Sales_Dataset.xlsx"

# Export the cleaned DataFrame to Excel.
# index=False prevents Pandas from creating an unnecessary
# index column in the Excel file.

df_cleaned.to_excel(
    output_path,
    index=False
)

print("Cleaned dataset saved successfully.")
print("File:", output_path)

Cleaned dataset saved successfully.
File: ../cleaned_data/ApexPlanet_Cleaned_Sales_Dataset.xlsx


In [72]:
# Read the newly created Excel file back into Pandas.
# This verifies that the exported file can be opened
# and loaded successfully.

df_saved = pd.read_excel(output_path)

print("Saved dataset shape:", df_saved.shape)
print("Saved dataset columns:")
print(df_saved.columns.tolist())

Saved dataset shape: (1000, 13)
Saved dataset columns:
['Transaction_ID', 'Order_ID', 'Order_Date', 'Customer_ID', 'Customer_Name', 'Age', 'Gender', 'City', 'Product', 'Category', 'Quantity', 'Unit_Price', 'Total_Sales']


In [73]:
# Compare the cleaned DataFrame with the dataset
# that was loaded back from the saved Excel file.

print(
    "Same number of rows:",
    len(df_cleaned) == len(df_saved)
)

print(
    "Same number of columns:",
    len(df_cleaned.columns) == len(df_saved.columns)
)

Same number of rows: True
Same number of columns: True


In [74]:
# Display the first five records from the saved dataset.
# This provides a final visual confirmation of the export.

df_saved.head()

,Transaction_ID,Order_ID,Order_Date,Customer_ID,Customer_Name,Age,Gender,City,Product,Category,Quantity,Unit_Price,Total_Sales
0,TXN000001,ORD100002,2025-02-25,CUST5529,Customer_227,30.0,Female,Bengaluru,Rice,Grocery,7,2829.77,19808.39
1,TXN000002,ORD100003,2025-10-14,CUST3127,Customer_182,63.0,Male,Bengaluru,Book,Education,5,27906.16,139530.80
2,TXN000003,ORD100004,2025-05-13,CUST8887,Customer_487,62.0,Female,Bengaluru,Book,Education,8,37491.06,299928.48
3,TXN000004,ORD100005,2025-12-02,CUST2515,Customer_470,65.0,Female,Kolkata,Mobile,Electronics,9,28541.36,256872.24
4,TXN000005,ORD100006,2025-11-20,CUST4796,Customer_380,44.0,Male,Bengaluru,Rice,Grocery,10,14036.59,140365.90


# 13. Final Validation of Cleaned Dataset

## Objective

The final exported dataset will be validated to ensure that
the saved file is consistent, complete, and ready for analysis.

We will verify:

- Dataset dimensions.
- Column structure.
- Data types.
- Unique Transaction_ID values.
- Missing values.
- Duplicate complete rows.
- Total_Sales business rule.
- Original Order_ID preservation.

In [75]:
# Check the number of rows and columns in the saved dataset.
# The cleaned dataset should contain 1,000 records and 13 columns.

print("Rows:", df_saved.shape[0])
print("Columns:", df_saved.shape[1])

# Display the column names.

print("\nColumns:")
print(df_saved.columns.tolist())

Rows: 1000
Columns: 13

Columns:
['Transaction_ID', 'Order_ID', 'Order_Date', 'Customer_ID', 'Customer_Name', 'Age', 'Gender', 'City', 'Product', 'Category', 'Quantity', 'Unit_Price', 'Total_Sales']


In [76]:
# Display the data type of every column in the saved dataset.
# This verifies that important columns were preserved correctly
# after exporting and reloading the Excel file.

df_saved.dtypes

Transaction_ID               str
Order_ID                     str
Order_Date        datetime64[us]
Customer_ID                  str
Customer_Name                str
Age                      float64
Gender                       str
City                         str
Product                      str
Category                     str
Quantity                   int64
Unit_Price               float64
Total_Sales              float64
dtype: object

In [77]:
# Verify that every transaction has a unique identifier
# and that no Transaction_ID values are missing.

print(
    "Unique Transaction_IDs:",
    df_saved["Transaction_ID"].nunique()
)

print(
    "Missing Transaction_IDs:",
    df_saved["Transaction_ID"].isna().sum()
)

print(
    "All Transaction_IDs unique:",
    df_saved["Transaction_ID"].is_unique
)

Unique Transaction_IDs: 1000
Missing Transaction_IDs: 0
All Transaction_IDs unique: True


In [78]:
# Check whether any complete duplicate records exist
# in the final saved dataset.

complete_duplicates = df_saved.duplicated().sum()

print(
    "Complete duplicate rows:",
    complete_duplicates
)

Complete duplicate rows: 0


In [79]:
# Display remaining missing values in the final dataset.
# These should correspond to the unresolved Age and City values
# documented during the cleaning process.

final_missing = df_saved.isna().sum()

final_missing[
    final_missing > 0
]

Age     19
City    11
dtype: int64

In [80]:
# Recalculate Total_Sales from Quantity and Unit_Price
# using the actual saved dataset.

saved_calculated_sales = (
    df_saved["Quantity"] *
    df_saved["Unit_Price"]
)

saved_sales_mismatch = df_saved[
    saved_calculated_sales.round(2) !=
    df_saved["Total_Sales"].round(2)
]

print(
    "Total_Sales formula mismatches:",
    len(saved_sales_mismatch)
)

Total_Sales formula mismatches: 0


In [81]:
# Confirm that the original Order_ID values were preserved.
# In particular, verify that the repeated ORD100050 values
# are still present rather than being deleted or overwritten.

print(
    "ORD100050 records:",
    (df_saved["Order_ID"] == "ORD100050").sum()
)

ORD100050 records: 9


In [82]:
# Create a compact final validation summary.

validation_summary = {
    "Rows": len(df_saved),
    "Columns": len(df_saved.columns),
    "Unique_Transaction_IDs": df_saved["Transaction_ID"].nunique(),
    "Complete_Duplicates": df_saved.duplicated().sum(),
    "Missing_Age": df_saved["Age"].isna().sum(),
    "Missing_City": df_saved["City"].isna().sum(),
    "Sales_Mismatches": len(saved_sales_mismatch),
    "Order_Date_Is_Datetime":
        pd.api.types.is_datetime64_any_dtype(
            df_saved["Order_Date"]
        ),
    "Transaction_IDs_Unique":
        df_saved["Transaction_ID"].is_unique
}

pd.Series(validation_summary)

Rows                      1000
Columns                     13
Unique_Transaction_IDs    1000
Complete_Duplicates          0
Missing_Age                 19
Missing_City                11
Sales_Mismatches             0
Order_Date_Is_Datetime    True
Transaction_IDs_Unique    True
dtype: object